designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,312 @@
+* 👍👍 Extends Existing API:  No Breaking API Changes, completely backwards compatible
+* 👍 No Nesting Required: Leaves budgets at the top level of the api.
+* 👎 Limited Generalization of Reason Controls: With reason being clearly tied to budgets, and other api logic being driven by disruption reason, we lose the chance to generalize per Reason controls. If we ever decide we need a place per action,  there will be some duplication for reason. 
+
+
+### Preferred Option: List Approach
+Given the comparison, the preferred design is the List Approach. It provides the necessary flexibility for managing multiple disruption reasons under a single budget, while reducing configuration complexity. This approach extends the existing API without introducing breaking changes and simplifies management for scenarios where multiple disruption reasons share similar constraints.
+
+### Approach: Defining Per Reason Controls  
+Ideally, we could move all generic controls that easily map into other reasons into one set of reason controls, this applies to budgets and other various disruption controls that could be more generic. 
+#### Example 
+
+```yaml 
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+metadata:
+  name: example-nodepool
+spec:
+  disruption:
+    defaults:
+      budgets: 
+        - nodes: 10% 
+          schedule: "0 0 1 * *"
+          duration: 1h 
+    consolidation:
+      consolidationPolicy: WhenUnderutilized
+      disruptAfter: "30m"
+    drift:
+      budgets:
+        - nodes: "20%"
+          schedule: "0 0 1 * *"
+          duration: "1h"
+        - nodes: "10%"
+          schedule: "0 0 * * 0"
+          duration: "2h"
+        - nodes: "50%" 
+          schedule: "@monthly"
+```
+#### Considerations 
+Some of the API choices for a given reason seem to follow a similar pattern. These include ConsolidateAfter, ExpireAfter, and there are discussions about introducing a global DisruptAfter. Moreover, when discussing disruption budgets, we talk about adding behavior for each reason. It appears there is a need for disruption controls within the budgets for each reason, not just overall.
+
+This approach aligns well with controls that apply to all existing reasons. The proposal presented here is similar to the one mentioned above in relation to the reasons we allow to be defined (underutilized, drifted, expired, empty).
+
+This proposal is currently scoped for disruptionBudgets by reason. However, we should also consider incorporating other generic disruption controls into the PerReasonControls, even if we do not implement them immediately. Moving ConsolidateAfter and ExpireAfter into the per-reason controls is a significant migration that requires careful planning and its own dedicated design. This proposal simply demonstrates a potential model that highlights the benefits of defining controls at a per-reason level of granularity.
+
+### Pros + Cons 
```

Comment:
The other problem that I have with this approach is that I think that this tries to group things together too early without having strong reasoning from users that we should do it this way. We actually got burned by this by grouping all of the disruption behavior into the same stanza. We're now thinking that we should undo some of this, since it's looking like some of this behavior should get pushed down into the NodeClaim API now

Reply:
I agree that defining PerReasonControls before we have a large set of controls needed for each reason is too aggressive and one of the main reasons I axed this approach.